### PR TITLE
feat: disaggregated computation — worker capabilities and prefill offloading

### PR DESCRIPTION
**Pull request by Gajesh2007**
## Summary

Implement disaggregated computation so every machine in the network has useful work to do, including small Macs (Air, 16GB) that currently can't run full inference efficiently. Worker capabilities are auto-detected from hardware: small machines become prefill workers (compute-bound prompt processing), while larger machines handle full inference (memory-bandwidth-bound decode). The coordinator routes work based on capabilities, preferring dedicated prefill workers to avoid consuming full-inference capacity.

## Linked issue

N/A — new feature

## Test plan

- [x] `go test ./...` — all existing tests pass (68s API, 0.5s registry, 0s protocol, etc.)
- [x] `go test -run "Disaggregated|WorkerCap|Prefill|DeriveWorker"` — 16 new tests pass
- [x] `gofmt -l` — all modified Go files are properly formatted
- [x] `cargo fmt` — Rust files formatted
- [x] `cargo check` — Rust changes are syntactically correct (full build requires macOS)
- [x] Verified backward compatibility: old providers without `worker_capabilities` get them auto-derived from hardware

## Components touched

- [x] coordinator (Go)
- [x] provider (Rust)
- [ ] console-ui (Next.js)
- [ ] image-bridge (Python)
- [ ] app (macOS Swift)
- [ ] enclave (Swift)
- [ ] infra / CI / release
- [ ] docs

## Protocol / interface changes

- [x] Yes — described above and matching side updated

### New message types:
- `prefill_request` (Coordinator → Provider): tells a prefill worker to process a prompt
- `prefill_complete` (Provider → Coordinator): reports prefill completion with token count

### Extended messages:
- `register`: now includes optional `worker_capabilities` field
- Both Go (`coordinator/internal/protocol/messages.go`) and Rust (`provider/src/protocol.rs`) updated in sync

### New types:
- `WorkerCapabilities`: `can_full_inference`, `can_prefill`, `max_model_size_gb`, `prefill_models`

### API changes:
- `GET /v1/stats` response now includes `disaggregated_compute` section with worker counts

No version bump or migration needed — all changes are additive and backward compatible.

## Notes for reviewers

**First-principles design:**
- Prefill (prompt processing) is compute-bound — GPU cores matter, not memory bandwidth
- Decode (token generation) is memory-bandwidth-bound — needs fast memory (Pro/Max/Ultra chips)
- Small machines (Air, 16GB, 100 GB/s) are terrible at decode but have decent compute cores
- Solution: classify machines by capability and route work accordingly

**Hardware classification logic:**
- `≤16GB + <150 GB/s bandwidth` → prefill-only (Air-class, M1/M2/M3/M4 base)
- `≥16GB + ≥150 GB/s` or `≥24GB` → full inference (Pro/Max/Ultra)
- `≥8GB` → can always do prefill work
- `<8GB` → excluded from all work

**Routing:**
- `ReservePrefillWorker()` prefers prefill-only workers first, falls back to full-inference providers
- This ensures small machines get utilized without stealing capacity from decode-capable machines

**Follow-ups needed:**
- Provider main loop should handle `CoordinatorEvent::PrefillRequest` and run actual prefill via vllm-mlx
- Consumer handler can optionally split long-prompt requests into prefill + decode phases
- Console UI can show disaggregated worker stats on the dashboard
- Billing for prefill work (currently only full inference is billed)

<div><a href="https://cursor.com/agents/bc-1a06c6e3-b550-4914-bfae-9a35e5da5802"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-1a06c6e3-b550-4914-bfae-9a35e5da5802"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

### CODE REVIEW

**Comment by hankbobtheresearchoor** [comment]
## Hermes Agent Review
**Verdict:** Minor suggestions

### ⚠️ Warning

1. **Silent drop on `CompleteCh` send failure** (`provider.go:417-419`): The `select` with `default` silently drops the prefill completion signal if `CompleteCh` is full or already consumed. In a disaggregated pipeline, a dropped signal means the decode phase never starts for that request — the consumer hangs until timeout. Consider logging when the default branch is hit, or using a larger buffer on `CompleteCh`.

2. **No timeout on prefill worker reservation** (`registry.go` `ReservePrefillWorker`): The prefill routing loop has no deadline. If all prefill workers are permanently busy, a consumer request could block indefinitely waiting for a prefill slot. The regular inference path has a queue timeout; the prefill path should have one too.

3. **Magic constants in capability derivation** (`deriveWorkerCapabilities`): The 16GB/150GBs/4GB-reserve thresholds are hardcoded. These are architecture-specific heuristics that may not generalize across non-Apple hardware. Consider making them configurable or at least named constants with documented rationale.

### 💡 Suggestion

4. **PrefillCompleteMessage lacks error/status field**: If a prefill worker fails (OOM, model load error), there is no way to report failure back to the coordinator. Consider adding an optional `error` field or a separate `PrefillFailedMessage` type so the coordinator can re-dispatch or fail fast.

5. **Missing prefill billing**: The PR description notes this as a follow-up, but `handlePrefillComplete` records job success and marks the worker idle without any cost tracking. If this merges before billing is implemented, free prefill work could be exploited. At minimum, add a TODO or metrics counter for prefills served.

6. **`PrefillRequestBody` reuses `ChatMessage` but prefill output format is underspecified**: The `PrefillCompleteMessage` only returns `prompt_tokens` and `duration_secs`. Where does the KV cache / logits data get transferred to the decode worker? The current protocol seems to only signal completion without actually passing the prefill output, making the disaggregated pipeline incomplete.

### ✅ Looks Good

7. **Backward compatibility**: Auto-deriving `WorkerCapabilities` from hardware when the field is absent on `RegisterMessage` is clean — old providers seamlessly get classified.

8. **Routing preference for prefill-only workers**: Preferring dedicated prefill workers over full-inference providers preserves decode capacity — correct design for this topology.

9. **Test coverage**: 16 new tests covering serialization, round-trip, worker selection, and exclusion logic is solid.

*Reviewed by Hermes Agent (hankbob)*
